### PR TITLE
Bump Akka HTTP 10.2.0

### PR DESCRIPTION
**Pull request by ignasi35**
Refs #866 

As part of this PR:

- I've enabled `-Xfatal-warnings`.
- replaced `AkkaSslConfig` with the new `akka-managment-pki` which is inspired by `akka-pki`. This new module could eventually be promoted to Akka code or Akka HTTP as part of https://github.com/akka/akka-http/issues/3772

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -102,6 +104,16 @@ lazy val `akka-management` = project
     mimaPreviousArtifactsSet
   )
 
+lazy val `akka-management-pki` = project
+  .in(file("management-pki"))
+  .enablePlugins(AutomateHeaderPlugin)
+  .settings(
+    name := "akka-management-pki",
+    libraryDependencies := Dependencies.ManagementPki,
+    // Don't enable mima until 1.1.1
```
Comment (raboof) [inline build.sbt:113]: can we make this an issue attached to the 1.1.1 milestone?

File: discovery-kubernetes-api/src/main/scala/akka/discovery/kubernetes/KubernetesApiServiceDiscovery.scala [hunk 2]
```diff
@@ -92,14 +100,15 @@ class KubernetesApiServiceDiscovery(implicit system: ActorSystem) extends Servic
...
-  private val httpsContext = http.createClientHttpsContext(httpsConfig)
+  private val sslContext = {
+  val km: Array[KeyManager] = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm).getKeyManagers
+  val tm: Array[TrustManager] = PemManagersProvider.buildTrustManagers(PemManagersProvider.loadCertificate(settings.apiCaPath))
+  val random: SecureRandom = new SecureRandom
+    val default = SSLContext.getDefault
+    default.init(km, tm, random)
+    default
+  }
+  private val clientSslContext:HttpsConnectionContext = ConnectionContext.httpsClient(sslContext)
```
Comment (raboof) [inline discovery-kubernetes-api/src/main/scala/akka/discovery/kubernetes/KubernetesApiServiceDiscovery.scala:111]:
```suggestion
  private val clientSslContext: HttpsConnectionContext = ConnectionContext.httpsClient(sslContext)
```

File: lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/internal/KubernetesApiImpl.scala [hunk 3]
```diff
@@ -37,13 +49,16 @@ import scala.util.control.NonFatal
...
-  private lazy val httpsContext = http.createClientHttpsContext(httpsConfig)
+  private val sslContext = {
+    val km: Array[KeyManager] = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm).getKeyManagers
+    val tm: Array[TrustManager] = PemManagersProvider.buildTrustManagers(PemManagersProvider.loadCertificate(settings.apiCaPath))
+    val random: SecureRandom = new SecureRandom
+    val default = SSLContext.getDefault
+    default.init(km, tm, random)
+    default
+  }
+  private val clientSslContext:HttpsConnectionContext = ConnectionContext.httpsClient(sslContext)
```
Comment (raboof) [inline lease-kubernetes/src/main/scala/akka/coordination/lease/kubernetes/internal/KubernetesApiImpl.scala:61]:
```suggestion
  private val clientSslContext: HttpsConnectionContext = ConnectionContext.httpsClient(sslContext)
```

File: project/Dependencies.scala [hunk 4]
```diff
@@ -13,10 +11,9 @@ object Dependencies {
   val AkkaVersion = "2.6.10"
   val AkkaBinaryVersion = "2.6"
   // Align the versions in integration-test/kubernetes-api-java/pom.xml
-  val AkkaHttp101 = "10.1.11"
   val AkkaHttp102 = "10.2.0"
-  val AkkaHttpVersion = if (CronBuild) AkkaHttp102 else AkkaHttp101
-  val AkkaHttpBinaryVersion = if (CronBuild) "10.2" else "10.1"
+  val AkkaHttpVersion = AkkaHttp102
```
Comment (raboof) [inline project/Dependencies.scala:15]: Not sure if the intermediate `AkkaHttp102` is still needed

File: management-pki/src/main/scala/akka/pki/kubernetes/PemManagersProvider.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Convenience methods to ease building an SSLContext from k8s-provided PEM files.
+ */
+// Duplicate from https://github.com/akka/akka/blob/31f654768f86db68f4c22daa2cbd0bae28fc1fad/akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/PemManagersProvider.scala#L35
+// Eventually that will be a bit more open and we can reuse the class from akka in akka-management.
+// See also https://github.com/akka/akka-http/issues/3772
+@ApiMayChange
+object PemManagersProvider {
+
+  def buildKeyManagers(privateKey: PrivateKey, cert: X509Certificate, cacert: Certificate): Array[KeyManager] = {
```
Comment (raboof) [inline management-pki/src/main/scala/akka/pki/kubernetes/PemManagersProvider.scala:36]: Maybe remove, if we're not using it anywhere (yet)?

**Comment by raboof** [comment]
also needs scalafmt

File: management/src/main/scala/akka/management/scaladsl/AkkaManagement.scala [hunk 6]
```diff
@@ -144,17 +144,14 @@ final class AkkaManagement(implicit private[akka] val system: ExtendedActorSyste
...
-            effectiveBindHostname,
-            effectiveBindPort,
-            connectionContext = connectionContext,
-            settings = ServerSettings(system).withRemoteAddressHeader(true)
-          )
+        val securedBuilder = effectiveProviderSettings.httpsConnectionContext match {
+          case Some(httpsContext) => baseBuilder.enableHttps(httpsContext)
+          case None => baseBuilder
+        }
+        val serverFutureBinding = securedBuilder.bindFlow(RouteResult.routeToFlow(combinedRoutes))
```
Comment (jrudolph) [inline management/src/main/scala/akka/management/scaladsl/AkkaManagement.scala:154]: Is there a reason, it's not just that:

```suggestion
        val serverFutureBinding = securedBuilder.bind(combinedRoutes)
```

File: management-pki/src/main/scala/akka/pki/kubernetes/PemManagersProvider.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import javax.net.ssl.TrustManagerFactory
+
+/**
+ * Convenience methods to ease building an SSLContext from k8s-provided PEM files.
+ */
+// Duplicate from https://github.com/akka/akka/blob/31f654768f86db68f4c22daa2cbd0bae28fc1fad/akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/PemManagersProvider.scala#L35
+// Eventually that will be a bit more open and we can reuse the class from akka in akka-management.
+// See also https://github.com/akka/akka-http/issues/3772
+@ApiMayChange
+object PemManagersProvider {
```
Comment (jrudolph) [inline management-pki/src/main/scala/akka/pki/kubernetes/PemManagersProvider.scala:34]: Is this class thought be used by users? If not maybe we should leave it private for now?

File: cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapRetryUnreachableContactPointIntegrationSpec.scala [hunk 8]
```diff
@@ -133,13 +140,12 @@ class ClusterBootstrapRetryUnreachableContactPointIntegrationSpec extends AnyWor
...
     "start listening with the http contact-points on 3 systems" in {
       def start(system: ActorSystem, contactPointPort: Int) = {
-        import system.dispatcher
         implicit val sys = system
 
         val bootstrap = ClusterBootstrap(system)
         val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
         bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
-        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+        Http().newServerAt("127.0.0.1", contactPointPort).bindFlow(RouteResult.routeToFlow(routes))
```
Comment (jrudolph) [inline cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapRetryUnreachableContactPointIntegrationSpec.scala:148]:
```suggestion
        Http().newServerAt("127.0.0.1", contactPointPort).bind(routes)
```

File: cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapIntegrationSpec.scala [hunk 9]
```diff
@@ -125,13 +132,12 @@ class ClusterBootstrapIntegrationSpec extends AnyWordSpecLike with Matchers {
...
     "start listening with the http contact-points on 3 systems" in {
       def start(system: ActorSystem, contactPointPort: Int) = {
-        import system.dispatcher
         implicit val sys = system
 
         val bootstrap = ClusterBootstrap(system)
         val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
         bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
-        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+        Http().newServerAt("127.0.0.1", contactPointPort).bindFlow(RouteResult.routeToFlow(routes))
```
Comment (jrudolph) [inline cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapIntegrationSpec.scala:140]:
```suggestion
        Http().newServerAt("127.0.0.1", contactPointPort).bind(routes)
```

File: cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapExistingSeedNodesSpec.scala [hunk 10]
```diff
@@ -107,13 +109,12 @@ class ClusterBootstrapExistingSeedNodesSpec(system: ActorSystem)
...
     "start listening with the http contact-points on all systems" in {
       def start(system: ActorSystem, contactPointPort: Int) = {
-        import system.dispatcher
         implicit val sys: ActorSystem = system
 
         val bootstrap = ClusterBootstrap(system)
         val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
         bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
-        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+        Http().newServerAt("127.0.0.1", contactPointPort).bindFlow(RouteResult.routeToFlow(routes))
```
Comment (jrudolph) [inline cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapExistingSeedNodesSpec.scala:117]:
```suggestion
        Http().newServerAt("127.0.0.1", contactPointPort).bind(routes)
```

File: cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapDiscoveryBackoffIntegrationSpec.scala [hunk 11]
```diff
@@ -154,13 +161,12 @@ class ClusterBootstrapDiscoveryBackoffIntegrationSpec
...
     "start listening with the http contact-points on 2 systems" in {
       def start(system: ActorSystem, contactPointPort: Int) = {
-        import system.dispatcher
         implicit val sys = system
 
         val bootstrap: ClusterBootstrap = ClusterBootstrap(system)
         val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
         bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
-        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+        Http().newServerAt("127.0.0.1", contactPointPort).bindFlow(RouteResult.routeToFlow(routes))
```
Comment (jrudolph) [inline cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapDiscoveryBackoffIntegrationSpec.scala:169]:
```suggestion
        Http().newServerAt("127.0.0.1", contactPointPort).bind(routes)
```

**Review by jrudolph — COMMENTED** [review]
A few suggestions below

File: build.sbt (hunk 1 above)
Comment (ignasi35) [reply build.sbt:113]: #896

File: management/src/main/scala/akka/management/scaladsl/AkkaManagement.scala (hunk 6 above)
Comment (ignasi35) [reply management/src/main/scala/akka/management/scaladsl/AkkaManagement.scala:154]: No reason. I followed the tips on the deprecation messages.

**Comment by ignasi35** [comment]
I wasn't too sure about making `PemManagersProvider` public, TBH. Given your comments I've rolled it back to `private[akka]`.

the most important tests in this build are the integration tests with K8s.

**Review by raboof — APPROVED** [review]
Nice!

**Pull request merged**